### PR TITLE
feat(ext/http): stream auto resp body compression

### LINKED ISSUE #13835: Streaming HTTP response body compression
Fixed length response bodies were done in https://github.com/denoland/deno/pull/13769

### PR DESCRIPTION
**Pull request by lucacasonato**
This commit adds support for auto response body compression for
streaming bodies.

Alternative to https://github.com/denoland/deno/tree/compression_stream
that uses `async_compression` and tokio's io traits.

cc @AaronO @satyarohith 

Fixes #13835

### CODE REVIEW

**Review by satyarohith — COMMENTED** [review]
We should add/update a test that decompresses the streamed response in addition to asserting the headers.

**Comment by lucacasonato** [comment]
@satyarohith Added the tests, and fixed two bugs in the process.

**Comment by littledivy** [comment]
How bad does this affect performance? Benchmark results of https://github.com/denoland/deno/pull/14328/files will be useful

File: ext/http/lib.rs [hunk 1]
```diff
@@ -545,55 +547,60 @@ async fn op_http_write_headers(
...
+    "Accept-Encoding".to_string()
+  };
+  builder = builder.header("vary", &vary_value);
+
+  let accepts_compression = matches!(
+    *stream.accept_encoding.borrow(),
+    Encoding::Brotli | Encoding::Gzip
+  );
+  let should_compress = body_compressible
+    && (matches!(data, Some(ref data) if data.len() > 20) || data.is_none())
```
Comment (satyarohith) [inline ext/http/lib.rs:576]: I don't understand the `data.is_none()` check. Can you explain?

**Comment by lucacasonato** [comment]
Main:

```
~/p/g/d/deno ❯❯❯ ./third_party/prebuilt/mac/wrk -t8 -c20 -d10s http://localhost:4500
Running 10s test @ http://localhost:4500
  8 threads and 20 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency    23.35ms    2.84ms  45.56ms   90.45%
    Req/Sec    85.67     11.59   101.00     60.89%
  6911 requests in 10.10s, 33.78GB read
Requests/sec:    684.43
Transfer/sec:      3.35GB
~/p/g/d/deno ❯❯❯ ./third_party/prebuilt/mac/wrk -t8 -c20 -d10s http://localhost:4500
Running 10s test @ http://localhost:4500
  8 threads and 20 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency    23.37ms    2.19ms  39.28ms   84.06%
    Req/Sec    85.59     10.75   101.00     65.47%
  6906 requests in 10.10s, 33.75GB read
Requests/sec:    683.76
Transfer/sec:      3.34GB
~/p/g/d/deno ❯❯❯ ./third_party/prebuilt/mac/wrk -t8 -c20 -d10s http://localhost:4500
Running 10s test @ http://localhost:4500
  8 threads and 20 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency    22.39ms    1.80ms  46.27ms   79.98%
    Req/Sec    89.47     10.17   101.00     55.94%
  7213 requests in 10.10s, 35.23GB read
Requests/sec:    714.19
Transfer/sec:      3.49GB
```

This branch:

```
~/p/g/d/deno ❯❯❯ ./third_party/prebuilt/mac/wrk -t8 -c20 -d10s http://localhost:4500
Running 10s test @ http://localhost:4500
  8 threads and 20 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency    35.54ms    1.78ms  60.39ms   92.31%
    Req/Sec    56.23      7.52    70.00     83.52%
  4538 requests in 10.10s, 22.22GB read
Requests/sec:    449.35
Transfer/sec:      2.20GB
~/p/g/d/deno ❯❯❯ ./third_party/prebuilt/mac/wrk -t8 -c20 -d10s http://localhost:4500
Running 10s test @ http://localhost:4500
  8 threads and 20 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency    35.56ms    1.22ms  50.26ms   82.34%
    Req/Sec    56.06      7.32    60.00     85.15%
  4530 requests in 10.10s, 22.23GB read
Requests/sec:    448.51
Transfer/sec:      2.20GB
~/p/g/d/deno ❯❯❯ ./third_party/prebuilt/mac/wrk -t8 -c20 -d10s http://localhost:4500
Running 10s test @ http://localhost:4500
  8 threads and 20 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency    35.63ms    1.89ms  65.07ms   96.02%
    Req/Sec    56.03      7.88    60.00     85.27%
  4527 requests in 10.10s, 22.18GB read
Requests/sec:    448.24
Transfer/sec:      2.20GB
```

**Pull request merged**